### PR TITLE
Move auth to src

### LINKED ISSUE #3954: login only happening on second attempt
Reproducible 100% on pirati.ca and squeet.me so assuming this affects all (?) instances.

Reproduce:
* click login icon (the door icon)
* fill out login information
* click "Login"

Currently:
Something happens, browser loads and then user finds themselves back in the previous view and not logged in. A second login always works as expected.

Expected:
Login should work on first attempt.

screencast (next 60 days): https://upload.disroot.org/r/bZagnwFX#r6sP0TSK0u64WEnJgShIImEAg1vRknMaW5884tMJNvA=

### PR DESCRIPTION
**Pull request by MrPetovan**
**Forewarning**: This potentially is a breaking change for anyone using the API through OAuth. I don't have a setup where I can test this case at the moment but I'm open to any suggestion. I just need to test the login.

---

This is a pretty significant PR that unfortunately combines multiple tangents to fix a seemingly simple issue.

At its core, though, the main change is that instead of checking for login credentials in any request, we only do it in the Login module which is now a class in src (see #4091).

We can still show the login form anywhere, but it only submits credentials to the login module before redirecting to the desired page. It is enough for most cases, however we lost the ability to keep POST parameters when we request user authentication which was only used in the API in the context of OAuth authentication.

### CODE REVIEW

**Comment by annando** [comment]
We are using POST in the API when uploading pictures. So have you checked with a client like Twidere if you still can create an item with a picture?

**Comment by MrPetovan** [comment]
If you're already authenticated, then it will still work. Like I said, the only broken use case is authentication with OAuth. Once you're authenticated, it will work as before, and if you aren't authenticated at all, it won't work as before.

**Comment by annando** [comment]
In the API you are authenticated at every single action. There are no sessions at all in the API. (See the concept of REST) The only reason that we are using the ```$_SESSION``` variable there is for being able to test the API via web browser.

**Comment by MrPetovan** [comment]
Ooh, I definitely didn't get that. This would definitely break all POST requests made using OAuth.

**Comment by annando** [comment]
I don't know if OAuth is working at all 😁

**Comment by MrPetovan** [comment]
Doesn't Twidere use OAuth? I don't have a personal Android device I could test on, but if you merge this PR, we can expect @Rudloff to post an issue if it doesn't work anymore 😇

**Comment by Rudloff** [comment]
I personnaly use basic authentication in Twidere 😊
(It also supports OAuth but I've never tried it.)

**Comment by AlfredSK** [comment]
Nope. If I remember correctly you can choose between 'Basic', 'OAuth' and 'OAuth2'.

**Comment by annando** [comment]
Question is how to proceed here. I wouldn't want to proceed here when it is very likely that it will break some things.

**Comment by annando** [comment]
For testing purposes you could create some test script. I would recommend this, since otherwise it would be hard for you to debug i.

**Comment by MrPetovan** [comment]
I’ve never written test scripts before so I have no idea where to start. I’ve sucessfully tested the web login because that’s what I personally use and if nobody we know uses OAuth/OpenId or knows if it even actually works we should remove support for them.

**Comment by annando** [comment]
I meant testing the picture upload in some way - because this is something that is really used.

And NTW: Why does it now makes problems with POST?

**Comment by MrPetovan** [comment]
Because in POST requests with OAuth, we were passing the POST parameters in the login form itself, that would then be submitted back to Friendica along with the credentials.

I’m not exactly sure it’s even POST, but that’s the only thing that would make sense to me. In any case this wouldn’t work for file upload anyway, so it is was working before, it should still work, and vice-versa, if it wasn’t working, I didn’t fix it!

**Comment by annando** [comment]
So you think that we should gave it a try? Would you work on a possible file upload problem before XMas? I will be away in the next few days and I guess I will upload many photos. So this should work for me. :-)

**Comment by MrPetovan** [comment]
Like I said, I can only test the web auth, I don’t have a personal Android device to test the API through a third-party client.

**Comment by annando** [comment]
Do you have any smartphone?

**Comment by MrPetovan** [comment]
I have an Apple iPhone, yes.

**Comment by MrPetovan** [comment]
I just tested my branch again and I still need to provide my credentials twice to log in on a multi-user node, this needs more love.

**Comment by annando** [comment]
Since it was working in master (I guess), you could try this: https://git-scm.com/docs/git-bisect

**Comment by AlfredSK** [comment]
If you tell me what to test I could try Twidere OAuth and OAuth2 on Android 8. Should I test the login? Or login and post stuff (text, photo, ...)?

**Comment by MrPetovan** [comment]
Yeah, the login, with OAuth you aren’t supposed to provide your credentials over and over again. Which makes me think that if Twidere provides auth credentials with each request then it won’t work anymore either.

**Comment by MrPetovan** [comment]
@annando Oooh, nifty, I may give it a try!

**Comment by MrPetovan** [comment]
After a very successful use of `git-bisect`, I found the root issue for the double login while still suppressing the annoying session write failed notices. This PR is back on, but it still needs to be tested against the API using a third-party client.

**Comment by annando** [comment]
Great that you found the origin of the problem! (And like always it is some very small change)

You could try this client: https://itunes.apple.com/de/app/shoyu-statusnet-gnusocial/id1059687793?mt=8

I don't know if it uploads pictures, but you can check if the api still works basically.

**Comment by MrPetovan** [comment]
Thanks, I’ll try!

**Comment by MrPetovan** [comment]
Using Shoyu on my way to the airport I can log in, consult my private messages and my profile but I can’t get my timeline or my mentions.

It’s a good start, I assume the missing timeline and mentions are unrelated to my work but I still need to test the photo upload.

**Comment by MrPetovan** [comment]
Well, posting works, posting with an image crashes the app 🤔

**Comment by annando** [comment]
So the question is if the picture upload crashes because of your changes or because the app is so bad :-)

**Comment by MrPetovan** [comment]
But it’s not using OAuth either so I don’t think it’s related. I’ll try on bare develop in 11 days, see you thwn!

**Comment by annando** [comment]
Happy holidays!

**Comment by AlfredSK** [comment]
I promised to test OAuth with Twidere. I failed. I cannot say if the code is malfunctioning or if I don't understand how it should work. Sorry. :'-/

**Comment by MrPetovan** [comment]
Did you test with this branch or the current develop or master?

**Comment by AlfredSK** [comment]
Oh. Yes. Current develop. Oops. :-/

**Comment by MrPetovan** [comment]
Okay, that’s one less hurdle for this PR to be merged. Now I just have to test the picture upload.

**Comment by MrPetovan** [comment]
Photo upload works normally on the site, I'll move to fix compatibility with the Shoyu app.

**Comment by annando** [comment]
So you now tested the API? Is there something else that is currently a blocker?

**Comment by MrPetovan** [comment]
No, Shoyu was a dead end, the API is correctly returning data, but Shoyu doesn't display it. OAuth doesn't currently work on develop so it needs to be fixed in a different PR as well.

**Comment by annando** [comment]
So I suggest that we accept this PR and possibly revert it, if it breaks stuff. What does @rabuzarus and @tobiasd think about that?

**Comment by annando** [comment]
I was just thinking about accepting this PR, since the two hadn't reacted and to see what happens. :-)

File: mod/dfrn_request.php [hunk 1]
```diff
@@ -47,100 +48,69 @@ function dfrn_request_init(App $a)
...
-						 */
-
-						notice( t("This introduction has already been accepted.") . EOL );
+					if (strlen($r[0]['dfrn-id'])) {
+						// We don't need to be here. It has already happened.
+						notice(t("This introduction has already been accepted.") . EOL);
 						return;
-					}
-					else
+					} else
```
Comment (tobiasd) [inline mod/dfrn_request.php]: shouldn't here be some curly braces for the block after the elso?

**Comment by rabuzarus** [comment]
Sry I don't have the time to review such big PR's. I agree with you: lets accept it and see what happens

**Comment by tobiasd** [comment]
As @annando and @rabuzarus are ok with it, I'll merge it now.

**Comment by annando** [comment]
BTW: Posting images via API does work.

**Pull request merged**